proteus/LinearSolvers.py

```diff
@@ -809,38 +790,48 @@ def getFp(self,output_matrix=False):
         self.Fsys_rowptr = rowptr.copy()
         self.Fsys_colind = colind.copy()
         self.Fsys_nzval = nzval.copy()
+        self.Fsys_nzval.fill(0.0)
+
         L_sizes = self.L.getSizes()
         nr = L_sizes[0][0]
         nc = L_sizes[1][0]
-        self.Fsys =SparseMat(nr,nc,
-                             self.Fsys_nzval.shape[0],
-                             self.Fsys_nzval,
-                             self.Fsys_colind,
-                             self.Fsys_rowptr)
+
+        self.Fsys = SparseMat(nr,nc,
+                              self.Fsys_nzval.shape[0],
+                              self.Fsys_nzval,
+                              self.Fsys_colind,
+                              self.Fsys_rowptr)
         self.L.pde.q[('df',0,0)][...,0] = self.L.pde.q[('u',1)]
         self.L.pde.q[('df',0,0)][...,1] = self.L.pde.q[('u',2)]
-        import pdb
  #        self.L.pde.q[('df',0,0)][...,2] = self.L.pde.q[('u',3)]
         self.L.pde.getSpatialJacobian(self.Fsys)#notice, switched  to spatial
         self.Fsys_petsc4py = self.L.duplicate()
+
         F_csr_rep_local = self.Fsys.getSubMatCSRrepresentation(0,L_sizes[0][0])
         self.Fsys_petsc4py.setValuesLocalCSR(F_csr_rep_local[0],
                                              F_csr_rep_local[1],
                                              F_csr_rep_local[2],
                                              p4pyPETSc.InsertMode.INSERT_VALUES)
+
         self.Fsys_petsc4py.assemblyBegin()
         self.Fsys_petsc4py.assemblyEnd()
-        self.Cp = self.Fsys_petsc4py.getSubMatrix(self.linear_smoother.isp,self.linear_smoother.isp)
+
+        self.Cp = self.Fsys_petsc4py.getSubMatrix(self.linear_smoother.isp,
+                                                  self.linear_smoother.isp)
+
```

Comment:
Since you're just trying to set up the advection term alone, you should be able to print `self.Cp` here (for a small mesh) and check directly if `Cp_ij = - int dot(u w_j ,grad  w_i) dV ` and that it is equal to a the proper block in Fsys.

An alternative way to set up Fp is just to set it up all at once by setting the advection, diffusion, and mass coefficients corresponding to the PCD preconditioner. I'm not sure what will be easier in the long run.